industrydive/django-site-metatags
# TECH-10121: /feeds/topics/whatever is broken if topic doesn't exist

[TECH-10121: /feeds/topics/whatever is broken if topic doesn't exist](https://industrydive.atlassian.net/browse/TECH-10121)

### Non-technical Context
When a user attempts to visit a `feeds/topic` link where the topic does not exist, the page will return a 500 error instead of the correct 404 page.

### Technical Context
#### Problem
The 500 error was being thrown from `django-site-metatags` when Django attempted to render the 404 page.

On a normal rendering of a `feeds/topic/<slug>`, such as `feeds/topic/learning` on HR Dive, Django will return an RSS feed. This does not inherit from `website_base.html`, so the `metatag` template tags are never called

However, when the topic does not exist, Django will attempt to render the 404 page. This does inherit from `website_base.html` and will call the `metatag` template tag.

The error begins when the template tag attempts to resolve the request path to a view name.

`metatag/templatetags/metatag_tags.py:L18`
```
url = resolve_to_name(request.path)
```

As the url pattern is a `RegexURLPattern`, the `_pattern_resolve_to_name` method is used. In this case, the url resolver is not named, and the callback, `ImageNewsTopicFeed` does not have a `_callbackstr` attribute. Therefore, the name is constructed using the callback's `__module__` and `__name__`.

The error is raised when accessing `self.callback.__name__` as the `self.callback` does not have a `__name__` attribute.

`metatag/resolve_to_name_snippet.py:L17-18`
```
...
        else:
            name = '%s.%s' % (self.callback.__module__, self.callback.__name__)
...
```

#### Solution
This ticket rewrites the `_pattern_resolve_to_name` method to check if the callback has `__module__` and `__name__` attributes. If it does not have a `__name__` attribute, it uses the `__class__.__name__` attribute if it exists.

If neither condition is fulfilled, we return an empty string.

I'm 95% sure it doesn't actually matter what we return here, as it seems to be trying to resolve a `request.path` to a view name, which is apparently supported? You could pass a view name in and have metatags rendered on any template that are rendered by that view? We don't do this, so we're always relying on the path, the second part of the `Q` here:

`metatag/templatetags/metatag_tags.py:L19`
```
meta = URLMetatags.objects.get(Q(url=url) | Q(url=request.path))
```

### Reproduction Steps
- Load divesite on `main`
- Start the server
- Visit `/feeds/topic/asdf`
- Observe a 500 error

### Manual Testing Instructions
We need to load the changes from this package into divesite. This is kinda difficult to do, but here's likely the easiest way
- Run `make`
- Check out this branch of this repo locally
- `cd django-site-metatags`
- `docker cp metatag/resolve_to_name_snippet.py divesite_app_1:/usr/local/lib/python3.8/site-packages/metatag/resolve_to_name_snippet.py`
- Visit `localhost:8000/feeds/topic/asdf`
- You should get a 404 page

Make sure I haven't broken working topics
- Visit `localhost:8000/feeds/topic/learning` on HR Dive settings
- You should see an RSS feed

Clean up
- Run `make` 

### QA Checklist

#### Developer

- [N/A] Code changes include a migration
    - [N/A] Migration modifies large table and has been run against full db to estimate completion time
    - [N/A] Migration modifies a small table and has only been run against slim db
    - [N/A] There are no conflicting migration leaf nodes


#### Reviewer

    - If there are few or minor issues, add inline GitHub comments
    - If there are many issues or it seems like the developer did not follow the style guide, leave a comment asking them to do so
    - [N/A] Ask for inline comments on unclear sections
    - [N/A] Technical documentation
- [x] There are no conflicting migration leaf nodes
